Delete empty paragraphs when backspacing

noisysocks opened the pull request:
This is a rough attempt to fix #3624.

When backspacing from an empty paragraph into a block that has no merge function (e.g. an image), the empty paragraph should be deleted.

Out with the old:

![old-behaviour](https://user-images.githubusercontent.com/612155/33414437-7765fcbc-d5e6-11e7-8e0a-ec1588d59aa1.gif)

In with the new:

![new-behaviour](https://user-images.githubusercontent.com/612155/33414448-799fa3f2-d5e6-11e7-8dd6-2eaee9a66b99.gif)

### How to test

1. Have multiple paragraphs.
1. Backspace to empty out a paragraph and backspace into previous one, verify the empty paragraph disappears.
1. Now insert an image, and backspace into image from the empty paragraph block, verify that the empty paragraph disappears.

noisysocks commented on editor/effects.js:
```diff
@@ -217,6 +218,9 @@ export default {
 		// Only focus the previous block if it's not mergeable
 		if ( ! blockType.merge ) {
 			dispatch( focusBlock( blockA.uid ) );
+			if ( blockB.name === 'core/paragraph' && ! blockB.attributes.content.length ) {
```
My instinct here is to invent a `isBlockEmpty` helper which we would use instead of this very very specific logic. 

I'm not sure, though, since it probably _is_ only paragraphs that we want this behaviour for...

Thoughts, @aduth?

gziolo replied:
We could also introduce a flag that marks all removable blocks like heading or list. 

`blockB.attributes.content.length` - is it safe to assume that content is there and it's a string?

gziolo replied:
And yes, you can at least introduce a local variable which better explains what is happening in this conditional statement :)

youknowriad replied:
Maybe we could introduce a ìsEmpty` property to the block API and if not defined the block is never empty. Only the block knows how to read its attributes.

noisysocks replied:
I added an `isEmpty` property and implemented it for `core/paragraph`!

aduth replied:
I don't love the `isEmpty` property (more to a general concern of limiting number of properties on the block API). It seems to me if Editable is smart enough to know that it should merge its own content into previous, it can also be smart enough to know that it has no content to merge and should just remove. That said, I also don't love an `onRemove` being added to Editable. Wondering if we could enhance one of the existing like `onReplace` to also encompass behavior of "replacing with nothing" ?

noisysocks replied:
That's a good idea @aduth. I moved my deletion logic to `Editable` in a0d193b7291da64bc9ce795f0f185e979b1bcbf9.

gziolo commented:
This `merge` property is also a bit strange, try the following:

Add an empty list and a quote and try to delete quote with backspace. It merges them together producing the list of 3 empty items:

![empty-block](https://user-images.githubusercontent.com/699132/33420118-b3b19b92-d5ad-11e7-9c80-321f7168e405.gif)

noisysocks committed:
Improve core/quote -> core/list conversion

Reduce the amount of times that empty <li>s are created when converting
a core/quote to a core/list.

The conversion logic is:

* 'Quote' / 'Citation' -> `<li>Quote</li><li>Citation</li>`
* '' / 'Citation' -> `<li></li><li>Citation</li>`
* 'Quote' / '' -> `<li>Quote</li>`
* '' / '' -> ``

noisysocks commented:
@gziolo it looks like that strangeness is caused by an unrelated bug in the `core/quote` → `core/list` transform. e5383ce473f4766671e59aa8572f265b966a24b8 ought to fix it.

jasmussen commented:
Seriously cool work. In my testing, behavior wise, this works exactly like it needs to! 👍 👍 from an experience point of view.

jorgefilipecosta commented on blocks/library/list/index.js:
```diff
@@ -118,8 +118,9 @@ registerBlockType( 'core/list', {
-						values: value.map( ( item, index ) => <li key={ index } >{ get( item, 'children.props.children', '' ) } </li> )
-							.concat( citation ? <li key="citation">{ citation }</li> : [] ),
+						values: value.map( item => get( item, 'children.props.children' ) ).concat( citation )
+							.filter( text => typeof text === 'string' )
```
This was a nice find that we are creating a list with an empty <li> and this affects some functionality 👍
Unfortunately, we cannot be that strict and force the text to be a string. For example, if we have a quote with items in bold they are ignored in the transformation because these items are not a string.
Also if the user adds a quote with a line then 3 empty lines and another line, I'm not certain if we should just transform to a list with two items or to a list with one item 3 empty items and one last item.
Maybe we can just add a special case that verifies if value and citation are both empty/unset and if yes we just return with a list with no items. e.g: `if ( ( ! value || ! value.length ) && ! citation ) {`.

noisysocks replied:
I've fiddled with the logic in e2257f5dc42dd5c34c1bcce99d255f5a7fecb187. There's a surprising amount of edge cases in this logic. 

The logic I opted for is:

* `'Quote' / 'Citation'` → `<li>Quote</li><li>Citation</li>`
* `'' / 'Citation'` → `<li></li><li>Citation</li>`
* `'Quote' / ''` → `<li>Quote</li>`
* `'' / ''` → ``

aduth commented:
A small thing that I really appreciate from this change: Clicking into the "Write your story" and hitting Backspace resets to remove the inserted paragraph block.

aduth commented on blocks/editable/index.js:
```diff
@@ -507,15 +507,27 @@ export default class Editable extends Component {
+			const isEmpty = dom.isEmpty( rootNode );
+			if ( this.props.onReplace && isEmpty ) {
+				// Remove this block if the editor is empty
+				this.props.onReplace( [] );
```
One odd thing about the logic flow here: If I render an `<Editable />` with `onReplace`, but not `onMerge`, I won't have this emptiness replace behavior, since this is only reached if `onMerge` is assigned, even though it's not called.

aduth commented:
I should note to my review, functionally it works great and I'm glad to see it handled in Editable entirely.

aduth commented:
Hmm, a potential issue is that some blocks may have multiple Editables, and one of them being empty is not a sufficient metric to determine that the block should be replaced. Example: Pressing backspace from a quote citation, when the quote itself has text, shouldn't delete the block.

mcsf reviewed (changes requested):
Nice work, _but_ it seems the replace behavior takes precedence over the merge one, leading to unexpected outcomes. Note how focus is lost when deleting into another paragraph:


| Before  | After |
| ------------- | ------------- |
| ![gutenberg-3732-merge-p](https://user-images.githubusercontent.com/150562/33616400-b6cc419e-d9d4-11e7-8733-154d6e6789e0.gif) | ![gutenberg-3732-replace-p](https://user-images.githubusercontent.com/150562/33616427-c3816ebe-d9d4-11e7-9cde-75392c603a01.gif) |

noisysocks commented:
> Hmm, a potential issue is that some blocks may have multiple Editables, and one of them being empty is not a sufficient metric to determine that the block should be replaced.

Good point... there's a lot of ~dragons~ subtleties here.

This approach _kind of_ works right now because in practice it's only `core/paragraph` that has both `onMerge` and `onReplace` passed to its `Editable`.

Unfortunately that means that the bug that this PR attempts to fix is only fixed for paragraph blocks and not quotes (which has `onMerge` but not `onReplace`):

![quote-bug](https://user-images.githubusercontent.com/612155/33645013-acf2f26c-da9c-11e7-9ca9-f16f50baf300.gif)

If we give `core/quote` an `onReplace`, we fix the problem:

![quote-bug-2](https://user-images.githubusercontent.com/612155/33645054-f6f41346-da9c-11e7-8861-eceed6d27188.gif)

**BUT** doing this means we now have the problem you describe where a quote that contains only a citation can be incorrectly deleted:

![quote-bug-3](https://user-images.githubusercontent.com/612155/33645140-64c1bf68-da9d-11e7-8149-e47bab90551f.gif)

---

Long story short, I think that patching `Editable` is unworkable. To fix this bug _in the general case_ we need the merge/deletion logic to read the block's entire state.

I have two ideas:

1) Go back to what we had before where each block defines an `isEmpty` method which the `MERGE_BLOCKS` effect can use to determine if that block should be deleted when being merged. I like the explicitness of this approach, but it _does_ mean extra API surface.
2) Same as (1) but instead of each block defining an `isEmpty` we build a method that compares each of the block's attributes against the block's default attributes. If no attribute differs from its default, then the block is empty and can be deleted when being merged. I think this might be our best bet.

aduth commented:
Could we add a new callback `onRemove` to `Editable` which is called as in the current implementation where we delete an empty Editable? Then, the block could decide if `onRemove` should be handled as `onReplace( [] )` (i.e. paragraph, or in other blocks when first checking that all other fields are empty as well).

noisysocks committed:
Delete empty paragraphs when backspacing

When backspacing from an empty paragraph into a block that has no merge
function (e.g. an image), the empty paragraph should be deleted.

Empty lists, headings and quotes are treated similarly.

noisysocks committed:
Improve core/paragraph -> core/list conversion

Prevent empty paragraphs from being converted to an empty <li>.

noisysocks commented:
> Could we add a new callback onRemove to Editable which is called as in the current implementation where we delete an empty Editable? 

Nice suggestion. I've done this and it seems to work pretty well.

I also fixed a similar bug to what @gziolo found where if one presses backspace in an empty _paragraph_ block that follows a list block, an empty `<li>` is created.

aduth commented on blocks/editable/index.js:
```diff
@@ -507,17 +507,34 @@ export default class Editable extends Component {
 			event.preventDefault();
 			event.stopImmediatePropagation();
+
+			return;
```
Is the `return;` necessary?

aduth commented on blocks/editable/index.js:
```diff
@@ -507,17 +507,34 @@ export default class Editable extends Component {
+			}
+
+			const isEmpty = dom.isEmpty( rootNode );
+			if ( isEmpty && this.props.onRemove ) {
```
Minor: If `dom.isEmpty` is non-trivial ([sorta appears so](https://github.com/tinymce/tinymce/blob/master/src/core/src/main/js/dom/Empty.js)), we could achieve a minor optimization here with conditional short-circuiting by reordering the condition to test the truthy `onRemove` first:

```js
if ( this.props.onRemove && dom.isEmpty( rootNode ) ) {
```

aduth reviewed (approved):
This seems to work quite nicely.

mtias commented:
Nice work here.
